grakn/service/Session/util/Communicator.py

```diff
@@ -0,0 +1,112 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import six
+from collections import deque
+from six.moves import queue
+from grakn.exception.GraknError import GraknError
+
+
+class IterationResolver(six.Iterator):
+    def __init__(self, communicator, request, is_last_response=lambda x: True):
+        self._request = request
+        self._is_last_response = is_last_response
+        self._communicator = communicator
+        self._ended = False
+        self._response_buffer = deque()
+        communicator._send_with_resolver(self, request)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            return self._response_buffer.popleft()
+        except IndexError:
+            if self._ended:
+                raise StopIteration()
+            return self._communicator._block_for_next(self)
+
+    def _buffer_response(self, response):
+        self._response_buffer.append(response)
+
+    def _end(self):
+        self._ended = False
+
+
+class Communicator(six.Iterator):
+    def __init__(self, grpc_stream_constructor):
+        self._request_queue = queue.Queue()
+        self._resolver_queue = deque()
+        self._response_iterator = grpc_stream_constructor(self)
+        self._closed = False
+
+    def __iter__(self):
+        return self
+
+    # Used by the GRPC stream to iterate requests as they arrive
+    def __next__(self):
+        request = self._request_queue.get(block=True)
+        if request is None:
+            raise StopIteration()
+        return request
+
+    # Put a request for GRPC to consume
+    def _send_with_resolver(self, resolver, request):
+        self._resolver_queue.append(resolver)
+        self._request_queue.put(request)
+
+    def _block_for_next(self, resolver):
+        if self._closed:
+            raise GraknError("This connection is closed")
+        while True:
+            current = None
+            try:
+                current = self._resolver_queue[0]
+                response = next(self._response_iterator)
+                if current._is_last_response(response):
+                    self._resolver_queue.popleft()
+                    current._ended = True
+                if current is resolver:
+                    return response
+                current._buffer_response(response)  # This result is for another
+
+            except Exception as e:
+                self._closed = True
+                if not current:
+                    raise GraknError("Internal client/protocol error, request/response pair not matched: {0}\n\n Ensure client version is compatible with server version.".format(e))
+                raise GraknError("Server/network error: {0}\n\n generated from request: {1}".format(e, current._request))
+
+    def iteration_request(self, request, is_last_response):
+        return IterationResolver(self, request, is_last_response)
```

Comment:
I think that this method shouldn't live in the `Communicator` - it obfuscates the flow of control and data. The callers of `iteration_request` can directly instantiate `IterationResolver(communicator, request, is_last_response` and we avoid having super-circular code to follow (iterator -> communicator.iteration_request -> IterationResolver -> communicator._send_with_resolver), it could look like (iterator -> IterationResolver -> communicator) instead...

the only thing with this is im not sure where to put `single_request` as it's used in multiple places... sure we can figure something out though :)

Reply:
Hmm, I'm not sure I prefer this.

The circularity is rather inherent since there are two "directions" that the messages come from:
A) pulled by requesting a response from an `IterationResolver`
B) pushed to an `IterationResolver` by the `Communicator` when a later resolver needs to pull a response

We could avoid any circular methods at all by putting all of the methods on `Communicator`, but that is also not necessarily nice to read (and the `IterationResolver` iterations methods are required so that it can be used as an iterator by the client).

The way I think of `IterationResolver` is as more of an implementation detail of `Communicator` and not a public class, since `Communicator` relies heavily on the hidden parts of `IterationResolver` to work correctly, they are collaborator classes at the very least but `Communicator` is the "owner". `IterationResolver` also happens to be the client interface to keep things simple, since that way there is only one object to worry about per request.

The fact that it is hard to read is really just purely inherent to the problem we are trying to solve, but we abstract it by making our API as simple and understandable to the user class (`TransactionService`) as possible:
* Single request calls return a response
* Multi request calls return an iterator

Changing that would obfuscate what we want the communicator to achieve as a user.

Reply:
As an additional point, whilst the circularity of the code flow is harder to follow, it makes sense to me that the only "public" methods are the request methods and the `__next__` method on the `IterationResolver`, so those are the only places you need to start.

Reply:
Ok - it was just unclear to me whether the `Communicator` or the `IterationResolver` has control in this setup.

the other solution to retain the simple API you've made is to introduce a small class/layer that does iteration_request() and single_request() but i'm not sure that would make the code any easier to read

Reply:
I don't mind the circularity at a concept, just push it down a bit further so that the first thing the communicator does isn't create a `IterationResolver` which just calls the communicator again - the entry point could just be one directional (and the cirularity occurs deeper down)